feat!: add Storybook 10 and Nuxt 4 support

akornmeier opened the pull request:
This pull request introduces full support for Storybook 10 and Nuxt 4, along with several breaking changes and enhancements to the Nuxt Storybook integration. The updates include ESM-only compatibility, revised documentation, major dependency upgrades, and improvements to the build process and developer experience.

**Storybook 10 and Nuxt 4 Support**

- Added full compatibility with Storybook 10's ESM-only architecture and Nuxt 3.x/4.x, including documentation updates and breaking change notices. [[1]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR3-R19) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L44-R55) [[3]](diffhunk://#diff-e04d9bf2a883cfa68cc34666d7fe9a52b93b9be25f986252b6c5ce4d99145c4bL13-R26) [[4]](diffhunk://#diff-bfbfbb5dfefd97e7ea568f8d3f054dca513f07157bb69c61f9c1fa8ff9b78f22L13-R25) [[5]](diffhunk://#diff-adfa337ce44dc2902621da20152a048dac41878cf3716dfc4cc56d03aa212a56L22-R22)

**Dependency and Example Updates**

- Upgraded all relevant dependencies and peer dependencies to support Storybook 10.x and Nuxt 4.x across the main module, examples, and documentation. This includes changes in `package.json`, example app `package.json` files, and workflow scripts. [[1]](diffhunk://#diff-6f977aee1fb8ef2360701ec0225959c01f5f1c0af7e2334ea638a78266349a25L15-R27) [[2]](diffhunk://#diff-156e31632655b3d3267a348634a05c9625965b5b7ade5b7f8f6271d3064cd0f9L15-R24) [[3]](diffhunk://#diff-0f8ffb6f82d2a180f9426495e2faf686abe5e5bb83ff7034eb442f0ab9f81b98L14-R22) [[4]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L51-R52) [[5]](diffhunk://#diff-56f5f5ffd060da32fab1f75b97868ff73fb58ba42de22988c6790e244ee7722eL28-R52) [[6]](diffhunk://#diff-56f5f5ffd060da32fab1f75b97868ff73fb58ba42de22988c6790e244ee7722eL47-R84) [[7]](diffhunk://#diff-3e103440521ada06efd263ae09b259e5507e4b8f7408308dc227621ad9efa31eR29-R32)

**Build and Module Improvements**

- Refactored the build configuration for the Nuxt module to use `unbuild`, re-exported Storybook preset/preview, and improved type exports for better ESM/CJS compatibility and developer experience. [[1]](diffhunk://#diff-aec1f6ee0ca820ea9eb60589212ad524bb87cae9b34e59df922b91337c4be7e8R3-R19) [[2]](diffhunk://#diff-56f5f5ffd060da32fab1f75b97868ff73fb58ba42de22988c6790e244ee7722eL28-R52) [[3]](diffhunk://#diff-56f5f5ffd060da32fab1f75b97868ff73fb58ba42de22988c6790e244ee7722eL47-R84) [[4]](diffhunk://#diff-7f936f18177f894d8888260ca38a9a4e97a6a3ae68e9de9a2fd69cead654ebfeR1-R2) [[5]](diffhunk://#diff-16e64880a111e498a957c29568f4078a742c7b6377a86c749d5a76d5939c653bR1-R2)

**Developer Experience and CI Enhancements**

- Improved the end-to-end workflow for local publishing and testing with Verdaccio, including more robust registry handling and project initialization steps.

**Breaking Changes and Migration Notes**

- Updated documentation and changelogs to highlight breaking changes (e.g., Node.js 20.19+ requirement, Storybook 10.x ESM-only), and provided guidance for users of previous Storybook/Nuxt versions. [[1]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR3-R19) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L44-R55) [[3]](diffhunk://#diff-bfbfbb5dfefd97e7ea568f8d3f054dca513f07157bb69c61f9c1fa8ff9b78f22L13-R25)

**Other Notable Changes**

- Deferred Storybook startup in the Nuxt module until the Nuxt HTTP server is ready for improved reliability.
- Updated example and showcase configurations for consistency with new Storybook/Nuxt versions and best practices. [[1]](diffhunk://#diff-b191c27ec1b51d2a0f080724e28d4969835e57ac5b8f35791cba7e1409020462R3-R6) [[2]](diffhunk://#diff-b191c27ec1b51d2a0f080724e28d4969835e57ac5b8f35791cba7e1409020462L13-L16) [[3]](diffhunk://#diff-7711b3144d2f44fb6636247cf5ea06439c9181219d982fdfdb0b7db3b7e7ef6bL21-R26) [[4]](diffhunk://#diff-f9923def8659fb2e00ddde5b6ba36c378cd4e1f3a849281b71e818d0636eb516L5) [[5]](diffhunk://#diff-6f977aee1fb8ef2360701ec0225959c01f5f1c0af7e2334ea638a78266349a25L2-R2)

These changes collectively modernize the Nuxt Storybook integration, ensure compatibility with the latest ecosystem releases, and improve the onboarding and usage experience for developers.

rpauls-scayle commented:
Cheers @akornmeier, thanks for taking on the upgrade. 🙏🏻 

Out of curiosity, are there particular technical reason why the upgrade to Storybook v10 needs to be accompanied by raising the minimum supported version to Nuxt v4?

While Nuxt v3 will reach EoL end of January 2026, a portion of projects might not yet have migrated to Nuxt 4. Raising the minimum supported Nuxt version would exclude those projects from benefiting of the Storybook v10 upgrade.

akornmeier commented:
> Cheers @akornmeier, thanks for taking on the upgrade. 🙏🏻
> 
> Out of curiosity, are there particular technical reason why the upgrade to Storybook v10 needs to be accompanied by raising the minimum supported version to Nuxt v4?
> 
> While Nuxt v3 will reach EoL end of January 2026, a portion of projects might not yet have migrated to Nuxt 4. Raising the minimum supported Nuxt version would exclude those projects from benefiting of the Storybook v10 upgrade.

Great question @rpauls-scayle, I will take a look and see if that minimum support can be relaxed back to Nuxt 3 and get back to you. Thanks for your insight!

akornmeier commented:
> > Cheers @akornmeier, thanks for taking on the upgrade. 🙏🏻
> > Out of curiosity, are there particular technical reason why the upgrade to Storybook v10 needs to be accompanied by raising the minimum supported version to Nuxt v4?
> > While Nuxt v3 will reach EoL end of January 2026, a portion of projects might not yet have migrated to Nuxt 4. Raising the minimum supported Nuxt version would exclude those projects from benefiting of the Storybook v10 upgrade.
> 
> Great question @rpauls-scayle, I will take a look and see if that minimum support can be relaxed back to Nuxt 3 and get back to you. Thanks for your insight!

@rpauls-scayle - Was able to confirm that Nuxt >= v3.18.1 is compatible with the new version of Storybook. I went ahead and adjusted the Peer dependencies and updated the README.md to reflect the supported versions. Good catch, thanks!

tobiasdiez commented on examples/showcase/.storybook/main.js:
```diff
@@ -1,18 +1,53 @@
-  framework: {
-    name: '@storybook-vue/nuxt',
-    options: {},
+  async viteFinal(config) {
```
If these steps are really necessary, then they should be handled by the plugin. The user shouldn't have the need to overwrite the viteFinal method.

tobiasdiez commented on examples/starter/.storybook/preview.ts:
```diff
@@ -1,6 +1,4 @@
-import type { Preview } from '@nuxtjs/storybook'
```
Why was that changed?

akornmeier replied:
I was normalizing the configs while chasing that render bug. I will revert these.

tobiasdiez commented on CHANGELOG.md:
```diff
@@ -1,5 +1,59 @@
+
+### 🚨 Breaking Changes
+
+- ⚠️ **Requires Nuxt 4.0.0+** - Dropped support for Nuxt 3.x
```
This is no longer true, right?

tobiasdiez commented on eslint.config.mjs:
```diff
@@ -1,5 +1,7 @@
 // @ts-check
 import { createConfigForNuxt } from '@nuxt/eslint-config/flat'
+// For more info, see https://github.com/storybookjs/eslint-plugin-storybook#configuration-flat-config-format
+import storybook from 'eslint-plugin-storybook'
```
This is not used (and I think doesn't make much sense for us)

akornmeier commented on examples/showcase/package.json:
```diff
@@ -12,18 +12,19 @@
-    "@storybook/addon-docs": "9.1.16",
-    "@storybook/addon-links": "9.1.16",
-    "storybook": "9.1.16"
+    "@nuxtjs/storybook": "workspace:*",
```
Good question, I will look into that. What are your thoughts on changing this to:

`"@nuxtjs/storybook: latest",`

That would keep the StackBlitz examples up to date w/o a maintenance burden.

akornmeier commented on CHANGELOG.md:
```diff
@@ -1,5 +1,59 @@
+
+### 📖 Documentation
+
+- Added technical notes about Vite optimization workarounds for ESM-only packages
```
❎ removed.

akornmeier commented on package.json:
```diff
@@ -83,5 +83,10 @@
+  "pnpm": {
+    "overrides": {
+      "@nuxtjs/storybook": "workspace:*"
+    }
```
Adds local support for our local package, while supporting the StackBlitz configs in the `examples/*` directories.

akornmeier commented:
@chakAs3 / @tobiasdiez - The PR is ready for review. I reverted the bundler change, made the edits Tobias suggested, and updated e2e.yml to pull local packages from Verdaccio. This allows the use of `@latest` in the StackBlitz examples and ensures the CI is always working from the latest changes in the packages.

akornmeier commented:
Thanks! Happy holidays.

akornmeier commented:
> Happy new year @akornmeier i was testting after merging this PR the playground does not run storybook server. although the examples run, but the first story rendering is so slow and the routing example is broken too could you check this please.

I found a few issues with the `/playground` example and got to the root of the issue (which is actually related to issue #993. I can tag that issue with my PR to fix. Should have it wrapped up in a few minutes.
